File: python/src/mrsf/discovery.py
"""MRSF Discovery — resolve sidecar file paths per §3.3.

Discovery order:
 1. Check for .mrsf.yaml at repo/workspace root → use sidecar_root if defined.
 2. Otherwise, co-located sidecar next to the Markdown file.
"""

from __future__ import annotations

import os
from pathlib import Path

from ruamel.yaml import YAML

from .types import MrsfConfig

_yaml = YAML()

CONFIG_FILENAME = ".mrsf.yaml"
SIDECAR_SUFFIX = ".review.yaml"
SIDECAR_SUFFIX_JSON = ".review.json"


def find_workspace_root(start_dir: str | Path) -> str:
    """Find workspace/repo root by walking up looking for .mrsf.yaml or .git."""
    d = Path(start_dir).resolve()
    while True:
        if (d / CONFIG_FILENAME).exists() or (d / ".git").exists():
            return str(d)
        parent = d.parent
        if parent == d:
            break
        d = parent
    return str(Path(start_dir).resolve())


def load_config(workspace_root: str, config_path: str | None = None) -> MrsfConfig | None:
    """Load and validate .mrsf.yaml config. Returns None if not found."""
    if config_path:
        cfg_path = Path(config_path).resolve()
    else:
        cfg_path = Path(workspace_root) / CONFIG_FILENAME

    if not cfg_path.exists():
        return None

    raw = cfg_path.read_text(encoding="utf-8")
    parsed = _yaml.load(raw)

    if not isinstance(parsed, dict):
        return None

    config = MrsfConfig()

    sr = parsed.get("sidecar_root")
    if isinstance(sr, str):
        if os.path.isabs(sr):
            raise ValueError(f'.mrsf.yaml: sidecar_root must be a relative path (got "{sr}")')
        if ".." in sr:
            raise ValueError(f'.mrsf.yaml: sidecar_root must not contain ".." (got "{sr}")')
        config.sidecar_root = sr

    return config


def discover_sidecar(
    document_path: str | Path,
    *,
    cwd: str | None = None,
    config_path: str | None = None,
) -> str:
    """Discover sidecar file path for a document. Returns absolute path."""
    effective_cwd = cwd or os.getcwd()
    workspace_root = find_workspace_root(effective_cwd)
    config = load_config(workspace_root, config_path)

    doc_path = Path(document_path)
    if doc_path.is_absolute():
        rel_doc = str(doc_path.relative_to(workspace_root))
    else:
        rel_doc = str(document_path)

    if config and config.sidecar_root:
        return str(Path(workspace_root) / config.sidecar_root / (rel_doc + SIDECAR_SUFFIX))

    return str(Path(workspace_root) / (rel_doc + SIDECAR_SUFFIX))


def sidecar_to_document(sidecar_path: str | Path) -> str:
    """Given a sidecar path, resolve the Markdown document path."""
    abs_path = str(Path(sidecar_path).resolve())

    if abs_path.endswith(SIDECAR_SUFFIX):
        return abs_path[: -len(SIDECAR_SUFFIX)]
    elif abs_path.endswith(SIDECAR_SUFFIX_JSON):
        return abs_path[: -len(SIDECAR_SUFFIX_JSON)]

    return abs_path


def discover_all_sidecars(dir_path: str | Path) -> list[str]:
    """Discover all sidecar files in a directory (recursive)."""
    p = Path(dir_path).resolve()

    if p.is_file():
        return [str(p)]

    results: list[str] = []

    def walk(d: Path) -> None:
        try:
            entries = sorted(d.iterdir())
        except PermissionError:
            return
        for entry in entries:
            if entry.is_dir():
                if entry.name in ("node_modules", ".git", "__pycache__", ".venv"):
                    continue
                walk(entry)
            elif entry.name.endswith(SIDECAR_SUFFIX) or entry.name.endswith(SIDECAR_SUFFIX_JSON):
                results.append(str(entry))

    walk(p)
    return results

File: cli/src/lib/discovery.ts
/**
 * MRSF Discovery — resolve sidecar file paths per §3.3.
 *
 * Discovery order:
 *  1. Check for .mrsf.yaml at repo/workspace root → use sidecar_root if defined.
 *  2. Otherwise, co-located sidecar next to the Markdown file.
 */

import { readFile } from "node:fs/promises";
import { existsSync } from "node:fs";
import path from "node:path";
import yaml from "js-yaml";
import type { MrsfConfig } from "./types.js";

const CONFIG_FILENAME = ".mrsf.yaml";
const SIDECAR_SUFFIX = ".review.yaml";
const SIDECAR_SUFFIX_JSON = ".review.json";

/**
 * Find the workspace / repo root by walking up from `startDir` looking for
 * `.mrsf.yaml` or `.git`.
 */
export function findWorkspaceRoot(startDir: string): string {
  let dir = path.resolve(startDir);
  const { root } = path.parse(dir);
  while (dir !== root) {
    if (
      existsSync(path.join(dir, CONFIG_FILENAME)) ||
      existsSync(path.join(dir, ".git"))
    ) {
      return dir;
    }
    dir = path.dirname(dir);
  }
  return path.resolve(startDir);
}

/**
 * Load and validate .mrsf.yaml config. Returns null if not found.
 */
export async function loadConfig(
  workspaceRoot: string,
  configPath?: string,
): Promise<MrsfConfig | null> {
  const cfgPath = configPath
    ? path.resolve(configPath)
    : path.join(workspaceRoot, CONFIG_FILENAME);

  if (!existsSync(cfgPath)) return null;

  const raw = await readFile(cfgPath, "utf-8");
  const parsed = yaml.load(raw, { schema: yaml.JSON_SCHEMA }) as Record<string, unknown> | null;

  if (!parsed || typeof parsed !== "object") return null;

  const config: MrsfConfig = {};

  if (typeof parsed.sidecar_root === "string") {
    const sr = parsed.sidecar_root;

    // Reject absolute paths
    if (path.isAbsolute(sr)) {
      throw new Error(
        `.mrsf.yaml: sidecar_root must be a relative path (got "${sr}")`,
      );
    }

    // Reject path traversal
    if (sr.includes("..")) {
      throw new Error(
        `.mrsf.yaml: sidecar_root must not contain ".." (got "${sr}")`,
      );
    }

    config.sidecar_root = sr;
  }

  return config;
}

/**
 * Given a Markdown document path (relative to workspace root), resolve the
 * sidecar file path according to §3.3 discovery order.
 *
 * Returns an absolute path to the sidecar.
 */
export async function discoverSidecar(
  documentPath: string,
  options: { cwd?: string; configPath?: string } = {},
): Promise<string> {
  const cwd = options.cwd ?? process.cwd();
  const workspaceRoot = findWorkspaceRoot(cwd);
  const config = await loadConfig(workspaceRoot, options.configPath);

  // Normalize to a workspace-relative path
  const relDoc = path.isAbsolute(documentPath)
    ? path.relative(workspaceRoot, documentPath)
    : documentPath;

  if (config?.sidecar_root) {
    // §3.2 — alternate sidecar location
    return path.join(workspaceRoot, config.sidecar_root, relDoc + SIDECAR_SUFFIX);
  }

  // §3.1 — co-located
  return path.join(workspaceRoot, relDoc + SIDECAR_SUFFIX);
}

/**
 * Given a sidecar file path, resolve the Markdown document path.
 * Strips .review.yaml or .review.json suffix.
 * Returns an absolute path to the document.
 */
export function sidecarToDocument(
  sidecarPath: string,
  options: { cwd?: string } = {},
): string {
  const abs = path.resolve(sidecarPath);

  if (abs.endsWith(SIDECAR_SUFFIX)) {
    return abs.slice(0, -SIDECAR_SUFFIX.length);
  } else if (abs.endsWith(SIDECAR_SUFFIX_JSON)) {
    return abs.slice(0, -SIDECAR_SUFFIX_JSON.length);
  }

  return abs;
}

/**
 * Discover all sidecar files in a directory (recursive).
 */
export async function discoverAllSidecars(
  dirPath: string,
): Promise<string[]> {
  const { readdir, stat } = await import("node:fs/promises");
  const results: string[] = [];

  async function walk(dir: string): Promise<void> {
    const entries = await readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (entry.name === "node_modules" || entry.name === ".git") continue;
        await walk(full);
      } else if (
        entry.name.endsWith(SIDECAR_SUFFIX) ||
        entry.name.endsWith(SIDECAR_SUFFIX_JSON)
      ) {
        results.push(full);
      }
    }
  }

  const s = await stat(dirPath);
  if (s.isFile()) {
    results.push(path.resolve(dirPath));
  } else {
    await walk(path.resolve(dirPath));
  }

  return results;
}
